tensorflow/contrib/timeseries/python/timeseries/head.py

```diff
@@ -0,0 +1,344 @@
+      loss=model_outputs.loss,
+      mode=estimator_lib.ModeKeys.EVAL,
+      eval_metric_ops=metrics,
+      predictions={})
+
+  def _predict_ops(self, features):
+    """Add ops for prediction to the graph."""
+    with variable_scope.variable_scope("model"):
+      prediction = self.model.predict(features=features)
+    prediction[feature_keys.PredictionResults.TIMES] = features[
+      feature_keys.PredictionFeatures.TIMES]
+    return estimator_lib.EstimatorSpec(
+      predictions=prediction, mode=estimator_lib.ModeKeys.PREDICT)
+
+  def _serving_ops(self, features):
+    with variable_scope.variable_scope("model"):
+      prediction_outputs = self.model.predict(features=features)
+    with variable_scope.variable_scope("model", reuse=True):
+      filtering_outputs = self.state_manager.define_loss(self.model, features,
+                                                         estimator_lib.ModeKeys.EVAL)
+    return estimator_lib.EstimatorSpec(
+      mode=estimator_lib.ModeKeys.PREDICT,
+      export_outputs={
+        feature_keys.SavedModelLabels.PREDICT:
+          export_lib.PredictOutput(prediction_outputs),
+        feature_keys.SavedModelLabels.FILTER:
+          export_lib.PredictOutput(
+            _state_to_dictionary(filtering_outputs.end_state))
+      },
+      # Likely unused, but it is necessary to return `predictions` to satisfy
+      # the Estimator's error checking.
+      predictions={})
+
+  def _convert_feature_to_tensor(self, name, value):
+    """Casts features to the correct dtype based on their name."""
+    if name in [
+      feature_keys.TrainEvalFeatures.TIMES,
+      feature_keys.PredictionFeatures.TIMES
+    ]:
+      return math_ops.cast(value, dtypes.int64)
+    if name == feature_keys.TrainEvalFeatures.VALUES:
+      return math_ops.cast(value, self.model.dtype)
+    if name == feature_keys.PredictionFeatures.STATE_TUPLE:
+      return value  # Correct dtypes are model-dependent
+    return ops.convert_to_tensor(value)
+
+  def _gather_state(self, features):
+    """Returns `features` with state packed, indicates if packing was done."""
+    prefixed_state_re = re.compile(r"^" + feature_keys.State.STATE_PREFIX +
+                                   r"_(\d+)$")
+    numbered_state = []
+    for key, tensor in features.items():
+      search_result = prefixed_state_re.search(key)
+      if search_result:
+        numbered_state.append((int(search_result.group(1)), key, tensor))
+    if not numbered_state:
+      return features, False
+    features = features.copy()
+    for _, key, _ in numbered_state:
+      del features[key]
+    numbered_state.sort(key=lambda number, *_: number)
+    features[feature_keys.State.STATE_TUPLE] = nest.pack_sequence_as(
+      structure=self.model.get_start_state(),
+      flat_sequence=[tensor for _, _, tensor in numbered_state])
+    return features, True
+
+  def create_estimator_spec(
+          self, features, mode, logits=None, labels=None, train_op_fn=None):
```

Comment:
One-line docstring like "Performs basic error checking and returns an EstimatorSpec."